KAFKA-18928: Support dynamically changing configs for a specific dynamic quorum controller

jiafu1115 opened the pull request:
Fix https://issues.apache.org/jira/browse/KAFKA-18928

Currently. It only support dynamically changing configs for a specific
**static** quorum controller.

### Test Environment
Dynamic quorum environment:
`controller.quorum.bootstrap.servers=ip1:port,ip2:port,ip3:port
`
Static quorum environment:
`controller.quorum.voters=id1@ip1:port,id2@ip2:port,id3@ip3:port
`
### Test command
`./kafka-configs.sh--bootstrap-controller ip:port --entity-type brokers
--entity-name 1001 --command-config /etc/kafka/admin_config.properties
--alter --add-config max.connections=10086`

Completed updating config for broker 1001.

### Verify command
`./kafka-configs.sh --bootstrap-controller ip:port --entity-type brokers
--entity-name 1001 --command-config /etc/kafka/admin_config.properties
--describe  --all |grep -E 'max.connections|node'`

max.connections=10086 sensitive=false
synonyms={DYNAMIC_BROKER_CONFIG:max.connections=100005,
DEFAULT_CONFIG:max.connections=2147483647} node.id=1001 sensitive=false
synonyms={STATIC_BROKER_CONFIG:node.id=1001}

### Test result
The issue on the dynamic quorum:
```
java.util.concurrent.ExecutionException:
org.apache.kafka.common.errors.BrokerIdNotRegisteredException: No node
with id 1001 found.
        at
java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
        at
java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
        at
org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:173)
        at
kafka.admin.ConfigCommand$.alterResourceConfig(ConfigCommand.scala:425)
        at
kafka.admin.ConfigCommand$.alterConfig(ConfigCommand.scala:192)
        at
kafka.admin.ConfigCommand$.processCommand(ConfigCommand.scala:165)
        at kafka.admin.ConfigCommand$.main(ConfigCommand.scala:87)
        at kafka.admin.ConfigCommand.main(ConfigCommand.scala)
Caused by:
org.apache.kafka.common.errors.BrokerIdNotRegisteredException: No node
with id 1001 found.
```
<img width="1715" height="205" alt="image"
src="https://github.com/user-attachments/assets/e0b096b4-2a1c-48dd-8aaf-0cf9855437df"
/>

### The result of the fix
<img width="1877" height="226" alt="image"
src="https://github.com/user-attachments/assets/b09ca3d5-f470-4f43-936c-f8deb209b406"
/>

Reviewers: Alyssa Huang <ahuang@confluent.io>, TaiJuWu
 <tjwu1217@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

jiafu1115 committed:
support dynamic controller voters' alter configure

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
Remove useless log

jiafu1115 committed:
refactor the code for check if node id registered.

Signed-off-by: stroller <fujian1115@gmail.com>

jiafu1115 committed:
add related unit test

Signed-off-by: stroller <fujian1115@gmail.com>

jiafu1115 committed:
update the related documents

Signed-off-by: stroller <fujian1115@gmail.com>

ahuang98 reviewed (commented):
can we add tests for the dynamic config update? (it would also be good to check that the other controllers are not picking up the config change, just the targeted controller)

also, I wasn't aware that dynamically changing a config for a single kraft controller (even just static) actually worked given that the documentation mentions it isn't supported. are we sure that the config is actually applied correctly?

jiafu1115 commented:
Test: Dynamic Controller: Only change one and verify if others also been changed.
<img width="1875" height="369" alt="image" src="https://github.com/user-attachments/assets/1eeeb88d-02a1-4bc4-a0d4-51a9d434308c" />

<img width="1327" height="51" alt="image" src="https://github.com/user-attachments/assets/5cbf297f-5503-4843-a1b1-9ef02b942134" />


_[2025-12-05 23:52:39,776] INFO [SocketServer listenerType=CONTROLLER, nodeId=1001] Updating broker-wide maxConnections: 123456789 (kafka.network.SocketServer)_

For static controller. The test result is similar . It will be workable without this code change. (I also thought it should not work according to the document ). 

@ahuang98  Thanks for your review. I attach this test according to your suggestion. Help to check again. Thanks

jiafu1115 committed:
improve the document to classify the controller work mode

ahuang98 commented:
@jiafu1115 thanks for the response and for testing it out manually! 

Good to see that the config isn't picked up by the other controllers. I wanted to clarify that my bigger concern (maybe uninformed concern since I'm not very familiar with config code) is that we are able to set the config for individual controllers, but they might not actually take effect. For instance, maybe kraft correctly creates a config record in the metadata log that broker id 1001 has this new config, but is controller 1001 actually able to apply this config (would it actually limit max connections at the new value)?

And, we probably need additional tests in the code for the dynamic config update (for instance, check out DynamicConfigChangeTest)

jiafu1115 commented:
@ahuang98 thanks for reviewing again. you can check this code lines and compare with the log I pasted in last comment (I copied it here)
<img width="1330" height="637" alt="image" src="https://github.com/user-attachments/assets/2b51c9f0-0977-4e59-82f0-72059de114a6" />
[2025-12-05 23:52:39,776] INFO [SocketServer listenerType=CONTROLLER, nodeId=1001] Updating broker-wide maxConnections: 123456789 (kafka.network.SocketServer)

For the unit test. I can take some time to check if we can add the extra one for it.
BTW:  the root cause why I don't add much more test cases at the beginning is that the change is very small which allow more capacity. It don't change any existed logic. But I test it manually is due to I need to make sure it can work end to end.

thanks

jiafu1115 commented:
org.apache.kafka.server.BootstrapControllersIntegrationTest#testIncrementalAlterConfigsByControllers the existed test to prove static quorum can work

chia7712 commented:
> And, we probably need additional tests in the code for the dynamic config update (for instance, check out DynamicConfigChangeTest)

I agree. We should create a helper test to validate all dynamic configurations on the controller. We must also cover both static quorum and dynamic quorum. `ClusterTest` supports static quorum.  The dynamic quorum tests could be implemented by `KafkaClusterTestKit`

jiafu1115 committed:
add test for dynamic quorum controller

Signed-off-by: jiafu1115 <fujian1115@gmail.com>

jiafu1115 commented:
@chia7712 @ahuang98 add the related core test. thanks !
FYI:
Before the code change:
<img width="1815" height="272" alt="image" src="https://github.com/user-attachments/assets/3f30ad69-aaaa-4f50-8034-a0706cdbfbfa" />

chia7712 commented:
@jiafu1115 could you please fix the conflicts?

jiafu1115 committed:
Merge branch 'trunk' into controller

jiafu1115 commented:
> @jiafu1115 could you please fix the conflicts?

@chia7712  done. thanks

apalan60 commented:
Thanks for the PR. I tested it locally with a dynamic voter setup, and it successfully updates a single controller’s config without any exceptions.

```bash 
./bin/kafka-configs.sh --bootstrap-controller localhost:9093 --entity-type brokers --entity-name 2 --describe --all | grep log.cleaner.threads
./bin/kafka-configs.sh --bootstrap-controller localhost:9093 --entity-type brokers --entity-name 2 --alter --add-config log.cleaner.threads=6
./bin/kafka-configs.sh --bootstrap-controller localhost:9093 --entity-type brokers --entity-name 2 --describe --all | grep log.cleaner.threads

log.cleaner.threads=1 sensitive=false synonyms={DEFAULT_CONFIG:log.cleaner.threads=1}
Completed updating config for broker 2.
log.cleaner.threads=6 sensitive=false synonyms={DYNAMIC_BROKER_CONFIG:log.cleaner.threads=6, DEFAULT_CONFIG:log.cleaner.threads=1}
```

jiafu1115 commented:
@apalan60 Thank you!

chia7712 commented on docs/documentation/zk2kraft.html:
```diff
@@ -210,7 +210,7 @@ <h3 class="anchor-heading">Dynamic Controller Configurations</h3>
     --add-config max.connections=10000
                 </code></pre><br/>
-                It is not currently possible to apply a dynamic configuration on only a single controller.
+                Prior to version 4.3, it was not possible to apply a dynamic configuration on only a single dynamic controller.
```
how about "Prior to version 4.3, dynamic configuration updates were not supported unless a static quorum was used"?

jiafu1115 replied:
@chia7712 based on follow tests' result, I think your propose make sense. 

There is no way to update all controllers at once. --entity-default only changes the default value and does not override values that have already been set.

Highlighting “only a single dynamic controller” may be misleading, since users might assume there is a way to update all controllers, which is not the case.

1  alter command should contain entity-name or --entity-default  in alter command
root@ip-10-20-1-107:/usr/local/kafka/bin# ./kafka-configs.sh --bootstrap-controller --bootstrap-controller 10.20.1.107:9889 --entity-type brokers --command-config /etc/kafka/admin_config.properties --alter --add-config max.connections=10086
**An entity-name or default entity must be specified with --alter of users, clients, brokers or ips**

2  --entity-default  is to update the default value: DYNAMIC_DEFAULT_BROKER_CONFIG instead of the all controller.
root@ip-10-20-1-107:/usr/local/kafka/bin# ./kafka-configs.sh --bootstrap-controller 10.20.1.107:9889 --entity-type brokers **--entity-default**    --command-config /etc/kafka/admin_config.properties --alter --add-config **max.connections=10086**
Completed updating default config for brokers in the cluster.
root@ip-10-20-1-107:/usr/local/kafka/bin# ./kafka-configs.sh --bootstrap-controller 10.20.1.107:9889 --entity-type brokers --entity-name 1001 --command-config /etc/kafka/admin_config.properties --describe  --all |grep -E 'max.connections|node'
max.connections=10086 sensitive=false synonyms={**DYNAMIC_DEFAULT_BROKER_CONFIG:max.connections=10086**, DEFAULT_CONFIG:max.connections=2147483647}

3 alert with  --entity-name  will override the default value. (note: after changed. if alter with --entity-default. it won't change the final value. it only change the default value)
root@ip-10-20-1-107:/usr/local/kafka/bin# ./kafka-configs.sh  --bootstrap-controller 10.20.1.107:9889 --entity-type brokers  **--entity-name 1001**    --command-config /etc/kafka/admin_config.properties --alter --add-config max.connections=**10000**
Completed updating config for broker 1001.
root@ip-10-20-1-107:/usr/local/kafka/bin# ./kafka-configs.sh --bootstrap-controller 10.20.1.107:9889 --entity-type brokers --entity-name 1001 --command-config /etc/kafka/admin_config.properties --describe  --all |grep -E 'max.connections|node'
max.connections=10000 sensitive=false synonyms={**DYNAMIC_BROKER_CONFIG:max.connections=10000**, **DYNAMIC_DEFAULT_BROKER_CONFIG:max.connections=10086**, DEFAULT_CONFIG:max.connections=2147483647}

jiafu1115 replied:
@chia7712 done with the change. thanks for your review!

jiafu1115 committed:
improve the document due to there is not method to change all controller at one time.

jiafu1115 commented:
Attach additional test results based on @chia7712 's proposed doc's change: 
(not directly related to code change; recorded here to avoid forgetting this logic.)

There is no any method to change all controllers in only one time. change with --entity-default only change the default value. it won't change the final value if the value had been set.

1  alter command should contain entity-name or --entity-default  in alter command
root@ip-10-20-1-107:/usr/local/kafka/bin# ./kafka-configs.sh --bootstrap-controller --bootstrap-controller 10.20.1.107:9889 --entity-type brokers --command-config /etc/kafka/admin_config.properties --alter --add-config max.connections=10086
**An entity-name or default entity must be specified with --alter of users, clients, brokers or ips**

2  --entity-default  is to update the default value: DYNAMIC_DEFAULT_BROKER_CONFIG instead of the all controller.
root@ip-10-20-1-107:/usr/local/kafka/bin# ./kafka-configs.sh --bootstrap-controller 10.20.1.107:9889 --entity-type brokers **--entity-default**    --command-config /etc/kafka/admin_config.properties --alter --add-config **max.connections=10086**
Completed updating default config for brokers in the cluster.
root@ip-10-20-1-107:/usr/local/kafka/bin# ./kafka-configs.sh --bootstrap-controller 10.20.1.107:9889 --entity-type brokers --entity-name 1001 --command-config /etc/kafka/admin_config.properties --describe  --all |grep -E 'max.connections|node'
max.connections=10086 sensitive=false synonyms={**DYNAMIC_DEFAULT_BROKER_CONFIG:max.connections=10086**, DEFAULT_CONFIG:max.connections=2147483647}

3 alert with  --entity-name  will override the default value. (note: after changed. if alter with --entity-default. it won't change the final value. it only change the default value)
root@ip-10-20-1-107:/usr/local/kafka/bin# ./kafka-configs.sh  --bootstrap-controller 10.20.1.107:9889 --entity-type brokers  **--entity-name 1001**    --command-config /etc/kafka/admin_config.properties --alter --add-config max.connections=**10000**
Completed updating config for broker 1001.
root@ip-10-20-1-107:/usr/local/kafka/bin# ./kafka-configs.sh --bootstrap-controller 10.20.1.107:9889 --entity-type brokers --entity-name 1001 --command-config /etc/kafka/admin_config.properties --describe  --all |grep -E 'max.connections|node'
max.connections=10000 sensitive=false synonyms={**DYNAMIC_BROKER_CONFIG:max.connections=10000**, **DYNAMIC_DEFAULT_BROKER_CONFIG:max.connections=10086**, DEFAULT_CONFIG:max.connections=2147483647}

4 --entity-name "" equal with --entity-default 
root@ip-10-20-1-107:/usr/local/kafka/bin# ./[kafka-configs.sh](https://kafka-configs.sh/)  --bootstrap-controller 10.20.1.107:9889 --entity-type brokers  --entity-name  ""   --command-config /etc/kafka/admin_config.properties --alter --add-config max.connections=10000
Completed updating default config for brokers in the cluster.

chia7712 commented:
@jiafu1115 please fix the docs

jiafu1115 committed:
Merge branch 'keepsynctotrunk' of https://github.com/jiafu1115/kafka into controller

# Conflicts:
#	docs/documentation/zk2kraft.html
#	docs/upgrade.html

jiafu1115 committed:
update the document

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
update the document

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
Update upgrade.md for version 4.3.0 changes

Clarified the support for dynamically changing configs for dynamic quorum controllers.

jiafu1115 commented:
> @jiafu1115 please fix the docs

@chia7712 done. thanks

jiafu1115 committed:
reorder the items to reduce fixing conflicts

jiafu1115 committed:
reduce useless changes

Added information about new configuration options for coordinators.

jiafu1115 committed:
reduce useless changes

jiafu1115 committed:
improve the documents

chia7712 commented on server/src/test/java/org/apache/kafka/server/BootstrapControllersIntegrationTest.java:
```diff
@@ -357,4 +357,9 @@ public void testDescribeConfigs(ClusterInstance clusterInstance) throws Exceptio
+
+    @ClusterTest(controllers = 1, standalone = true)
+    public void testIncrementalAlterConfigsByControllersWithDynamicQuorum(ClusterInstance clusterInstance) throws Exception {
+        testIncrementalAlterConfigs(clusterInstance, true);
```
Could you add a test to ensure all controllers can accept dynamic configurations? The current logic only verifies the single initial controller

jiafu1115 replied:
@chia7712  done. add one test: create one 3 controller cluster and one admin client. then change 3 controller.

jiafu1115 committed:
add more unit test

Signed-off-by: jiafu1115 <fujian1115@gmail.com>

chia7712 commented on server/src/test/java/org/apache/kafka/server/BootstrapControllersIntegrationTest.java:
```diff
@@ -226,27 +226,29 @@ public void testIncrementalAlterConfigs(ClusterInstance clusterInstance) throws
     private void testIncrementalAlterConfigs(ClusterInstance clusterInstance, boolean usingBootstrapControllers) throws Exception {
+        Collection<Integer> nodeIds = usingBootstrapControllers ?
+                clusterInstance.controllerIds() :
+                List.of(clusterInstance.brokers().values().iterator().next().config().nodeId());
```
```java
        Collection<Integer> nodeIds = usingBootstrapControllers ?
                clusterInstance.controllerIds() :
                clusterInstance.brokers().keySet();
```

jiafu1115 replied:
@chia7712 done. thanks
<img width="806" height="220" alt="image" src="https://github.com/user-attachments/assets/ef44a76d-9518-4379-a929-e08a50e92f17" />

chia7712 commented on server/src/test/java/org/apache/kafka/server/BootstrapControllersIntegrationTest.java:
```diff
@@ -226,27 +226,29 @@ public void testIncrementalAlterConfigs(ClusterInstance clusterInstance) throws
+                    ConfigEntry entry = config.entries().stream().
+                            filter(e -> e.name().equals("my.custom.config")).
+                            findFirst().orElseThrow();
+                    assertEquals(DYNAMIC_BROKER_CONFIG, entry.source(),
```
why we don't verify the value?

jiafu1115 replied:
@chia7712 Done. I’ve already updated the test to use max.connections instead of my.custom.config, since the latter is treated as sensitive and its value cannot be queried. That's the reason for why existed tests don't check the value.

chia7712 commented on server/src/test/java/org/apache/kafka/server/BootstrapControllersIntegrationTest.java:
```diff
@@ -226,27 +226,29 @@ public void testIncrementalAlterConfigs(ClusterInstance clusterInstance) throws
+                ConfigResource defaultResource = new ConfigResource(BROKER, "");
+                Map<ConfigResource, Collection<AlterConfigOp>> alterations = Map.of(
+                        nodeResource, List.of(new AlterConfigOp(new ConfigEntry("my.custom.config", "foo"), AlterConfigOp.OpType.SET)),
+                        defaultResource, List.of(new AlterConfigOp(new ConfigEntry("my.custom.config", "bar"), AlterConfigOp.OpType.SET))
```
Maybe we should use a separate key for the cluster-level config

jiafu1115 replied:
@chia7712 Done. I add two configure items for test instead of only one.

jiafu1115 committed:
fix the test code for nodeId list

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
check the test value

Signed-off-by: Jian <fujian1115@gmail.com>

TaiJuWu commented on test-common/test-common-runtime/src/main/java/org/apache/kafka/common/test/junit/RaftClusterInvocationContext.java:
```diff
@@ -324,6 +324,9 @@ public void format() throws Exception {
                 clusterConfig.serverProperties().forEach(builder::setConfigProp);
+                if (clusterConfig.standalone()) {
+                    builder.setStandalone(true);
+                }
```
How about `builder.setStandalone(clusterConfig.standalone())` ?

jiafu1115 replied:
@TaiJuWu  Done. good catch. thanks!

jiafu1115 committed:
refactor the code to remove useless judge

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
add more test

Signed-off-by: Jian <fujian1115@gmail.com>

chia7712 commented on server/src/test/java/org/apache/kafka/server/BootstrapControllersIntegrationTest.java:
```diff
@@ -226,30 +227,61 @@ public void testIncrementalAlterConfigs(ClusterInstance clusterInstance) throws
+                verifyConfigValue(admin, defaultResource, SocketServerConfigs.MAX_CONNECTIONS_CONFIG,
+                        DYNAMIC_DEFAULT_BROKER_CONFIG, defaultMaxConnectionsValue);
+                verifyConfigValue(admin, defaultResource, SocketServerConfigs.MAX_CONNECTION_CREATION_RATE_CONFIG,
+                        DYNAMIC_DEFAULT_BROKER_CONFIG, defaultConnectionRateValue);
```
Would you mind verifying `SocketServer` to ensure the max connectio limit gets updated?

chia7712 replied:
It also addresses @ahuang98's comment (https://github.com/apache/kafka/pull/20938#pullrequestreview-3546706656)

jiafu1115 replied:
@chia7712 Thanks for review. Done. PTAL.

I added two extra checkpoints with https://github.com/apache/kafka/pull/20938/changes#diff-38ac8a8b4162b71f7b0a8244bd9d5f2ddb354b691cbd0b7699f77a5920f331a6R278 in the test.

BTW: https://github.com/apache/kafka/pull/20938#issuecomment-3634864330 is the real end to end test result to solve @ahuang98 's concern.

chia7712 replied:
Thanks for the updates. I wonder if the controller actually applies the latest dynamic configs, rather than just generating the config records

jiafu1115 replied:
make sense

jiafu1115 committed:
add more test checkpoint

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
fix merge code issue

Signed-off-by: Jian <fujian1115@gmail.com>

chia7712 commented on server/src/test/java/org/apache/kafka/server/BootstrapControllersIntegrationTest.java:
```diff
@@ -225,31 +228,93 @@ public void testIncrementalAlterConfigs(ClusterInstance clusterInstance) throws
         }
     }
 
+    private void verifySocketServerMaxConnectionsUpdated(Object broker, int expectedMaxConnections) throws Exception {
```
`broker` -> `node`

jiafu1115 replied:
@chia7712 done. thanks

chia7712 commented on server/src/test/java/org/apache/kafka/server/BootstrapControllersIntegrationTest.java:
```diff
@@ -225,31 +228,93 @@ public void testIncrementalAlterConfigs(ClusterInstance clusterInstance) throws
         testIncrementalAlterConfigs(clusterInstance, false);
     }
 
+    @SuppressWarnings("unchecked")
```
it is redundant

jiafu1115 replied:
@chia7712 done. thanks

chia7712 reviewed (commented):
@jiafu1115 thanks for this patch. overall LGTM. Two small comments remain

jiafu1115 committed:
improve the unit test according to code review

Signed-off-by: stroller <fujian1115@gmail.com>
